### DISCUSSION TITLE
Is it bad to have your service key as a vault secret and envoke an edge function with it as a auth header?

### DISCUSSION THREAD

**Discussion by cwooldridge1** [supabase/supabase#21791] [Questions] [answered]
I have a pg_cron function that is used to clean up unreferenced storage objects from storage (this is caused be things like a user uploading a file but never actually saving what they where doing). This cron function envokes an edge function with the auth header using the service key. Not sure if this is an anti pattern or if there is a better way to do it.

My alternative solution was just operating as the service role from the edge function and checking to see if a passkey was sent to ensure non-authorized users couldn't use this function as of my understanding there is no such thing as a private edge function.

Just for reference this is how I have been using it:

```sql
create function private.supabase_service_key()
returns text
language plpgsql
security definer
as $$
declare
  secret_value text;
begin
  select decrypted_secret into secret_value from vault.decrypted_secrets where name = 'supabase_service_key';
  return secret_value;
end;

$$;
select
  cron.schedule(
    'storage-clean-up',
    '0 0 * * *', -- every day
    $$
    select
      net.http_post(
          url := supabase_url() || '/functions/v1/cron-storage-cleanup',
          headers := jsonb_build_object(
              'Content-Type', 'application/json',
              'Authorization', 'Bearer ' || private.supabase_service_key()
          ),
          body := jsonb_build_object(
            'paths', array(select name from private.get_images_to_delete())
          )
      ) as request_id;
    $$
  );
```

**Comment by GaryAustin1** [comment, marked as answer]
No, that is the way most are going until Supabase provides those keys to Postgres SQL.  You have 3 choices to get it in SQL code, hardcode, use a private schema or use Vault.

**Reply by cwooldridge1** [reply]
Thanks! I just check out your storage helper repo https://github.com/GaryAustin1/supa-file-helper looks useful. I was going to go with this approach but it just seemed easier to envoke an edge function with the files I wanted to delete.

**Comment by obenguerah** [comment]
> ```sql
> supabase_url()
> ```

Hi @cwooldridge1 , wondering how do you get this function? (ie: read supabase url from a cron/sql)

**Reply by cwooldridge1** [reply]
Here is the function I defined for that:
```
create function supabase_url()
returns text
language plpgsql
security definer
as $$
declare
  secret_value text;
begin
  select decrypted_secret into secret_value from vault.decrypted_secrets where name = 'supabase_url';
  return secret_value;
end;
$$;
```
so in your `seed.sql` or in the ui you will have to make a vault secret aswell of: 
select vault.create_secret(
  'http://api.supabase.internal:8000',
  'supabase_url'
);
